coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/CalciteTrinoUDFOperatorTransformers.java

```diff
@@ -0,0 +1,220 @@
+  }
+
+  private static void addDaliFunctionId(Set<String> daliFunctionIdSet, String daliFunctionName, int operandNum) {
+    daliFunctionIdSet.add(daliFunctionName + "_" + operandNum);
+  }
+
+  private static void addDaliUDFs(List<SqlCallTransformer> operatorTransformers, Set<String> daliFunctionIdSet) {
+    ImmutableMultimap<String, Function> registry = HIVE_REGISTRY.getRegistry();
+    Converter<String, String> caseConverter = CaseFormat.UPPER_CAMEL.converterTo(CaseFormat.LOWER_UNDERSCORE);
+    for (Map.Entry<String, Function> entry : registry.entries()) {
+      // we cannot use entry.getKey() as function name directly, because keys are all lowercase, which will
+      // fail to be converted to lowercase with underscore correctly
+      final String hiveFunctionName = entry.getValue().getFunctionName();
+      if (!hiveFunctionName.startsWith("com.linkedin")) {
+        continue;
+      }
+      String[] nameSplit = hiveFunctionName.split("\\.");
+      // filter above guarantees we've at least 2 entries
+      String className = nameSplit[nameSplit.length - 1];
+      String funcName = caseConverter.convert(className);
+      SqlOperator op = entry.getValue().getSqlOperator();
+      for (int i = op.getOperandCountRange().getMin(); i <= op.getOperandCountRange().getMax(); i++) {
+        if (!daliFunctionIdSet.contains(hiveFunctionName + "_" + i)) {
+          operatorTransformers.add(createOperatorTransformer(op, i, funcName));
+        }
+      }
+    }
+  }
+
+  private static SqlOperator daliToCalciteOp(String className) {
+    return HIVE_REGISTRY.lookup(className).iterator().next().getSqlOperator();
+  }
+
+  private static SqlCallTransformer createOperatorTransformer(SqlOperator calciteOp, int numOperands,
+      String trinoUDFName) {
+    return createOperatorTransformer(calciteOp, numOperands,
+        UDFMapUtils.createUDF(trinoUDFName, calciteOp.getReturnTypeInference()));
+  }
+
+  private static SqlCallTransformer createOperatorTransformer(SqlOperator calciteOp, int numOperands,
+      String trinoUDFName, String operandTransformer, String resultTransformer) {
+    return createOperatorTransformer(calciteOp, numOperands,
+        UDFMapUtils.createUDF(trinoUDFName, calciteOp.getReturnTypeInference()), operandTransformer, resultTransformer);
+  }
+
+  private static SqlCallTransformer createOperatorTransformer(SqlOperator calciteOp, int numOperands,
+      SqlOperator trinoOp) {
+    return createOperatorTransformer(calciteOp, numOperands, trinoOp, null, null, null);
+  }
+
+  private static SqlCallTransformer createOperatorTransformer(SqlOperator calciteOp, int numOperands,
+      SqlOperator trinoOp, String operandTransformer, String resultTransformer) {
+    return createOperatorTransformer(calciteOp, numOperands, trinoOp, operandTransformer, resultTransformer, null);
+  }
+
+  private static SqlCallTransformer createOperatorTransformer(SqlOperator calciteOp, int numOperands,
```

Comment:
I feel that we don't need so many helper methods? May remove some of them if possible